src/icq/system.rs

```diff
@@ -53,27 +53,65 @@ impl ICQSystem {
             match purple_message {
                 PurpleMessage::Login(account_info) => self.login(account_info).await,
             }
-            self.ping().await
         }
     }
 
-    async fn ping(&mut self) {
-        self.tx.send(SystemMessage::Ping).await;
-    }
-
     async fn login(&mut self, account_info: AccountInfo) {
         log::debug!("login");
-        match protocol::register(&account_info.phone_number, || {
-            log::debug!("read_code");
-            self.read_code(&account_info.account)
-        })
-        .await
-        {
-            Ok(_) => (),
-            Err(error) => {
-                log::error!("Failed to register account: {:?}", error);
+        let mut registered_account_info = {
+            account_info
+                .account
+                .proxy(&mut self.tx)
+                .exec(|account| {
+                    let token = account.get_string("token", "");
```

Comment:
Didn't you create serializers and deseriailizers for this?

Maybe atleast use enums?